src/coreclr/jit/lower.cpp

```diff
@@ -5753,6 +5753,26 @@ void Lowering::LowerShift(GenTreeOp* shift)
         shift->gtOp2->ClearContained();
     }
     ContainCheckShiftRotate(shift);
+
+// Check if we can fold e.g. '(ulong/long)x << 2' where x is int/uint to ubfiz/sbfiz
+#ifdef TARGET_ARM64
+    if (comp->opts.OptimizationEnabled() && shift->OperIs(GT_LSH) && shift->gtGetOp1()->OperIs(GT_CAST) &&
+        shift->gtGetOp2()->IsCnsIntOrI() && !shift->isContained() && varTypeIsIntegral(shift))
+    {
+        GenTreeIntCon* cns  = shift->gtGetOp2()->AsIntCon();
+        GenTreeCast*   cast = shift->gtGetOp1()->AsCast();
+
+        if (!cast->isContained() && !cast->IsRegOptional() && varTypeIsLong(shift) && !cast->CastOp()->isContained() &&
+            (varTypeToSigned(cast->CastFromType()) == TYP_INT) && (cns->IconValue() > 0) && (cns->IconValue() < 32) &&
```

Comment:
Why exclude small-typed nodes? They produce `TYP_INT` too (with one very special exception that exists only on x86). I think this could be just `genActualTypeIsInt(cast->CastOp())`.

(Applies to the assert in codegen too)

Reply:
Thanks for feedback 🙂 , but I just wanted to see green CI with these conservative checks and i64 <- i32 case only. Now I can extend it for small integers and drop some redundant checks